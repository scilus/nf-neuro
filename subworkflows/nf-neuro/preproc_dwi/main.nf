include { DENOISING_MPPCA as DENOISE_DWI } from '../../../modules/nf-neuro/denoising/mppca/main'
include { DENOISING_MPPCA as DENOISE_REVDWI } from '../../../modules/nf-neuro/denoising/mppca/main'
include { BETCROP_FSLBETCROP } from '../../../modules/nf-neuro/betcrop/fslbetcrop/main'
include { IMAGE_CROPVOLUME } from '../../../modules/nf-neuro/image/cropvolume/main'
include { PREPROC_N4 as N4_DWI } from '../../../modules/nf-neuro/preproc/n4/main'
include { PREPROC_NORMALIZE as NORMALIZE_DWI } from '../../../modules/nf-neuro/preproc/normalize/main'
include { IMAGE_RESAMPLE as RESAMPLE_DWI } from '../../../modules/nf-neuro/image/resample/main'
include { IMAGE_RESAMPLE as RESAMPLE_MASK } from '../../../modules/nf-neuro/image/resample/main'
include { UTILS_EXTRACTB0 as EXTRACTB0_RESAMPLE } from '../../../modules/nf-neuro/utils/extractb0/main'
include { UTILS_EXTRACTB0 as EXTRACTB0_TOPUP } from '../../../modules/nf-neuro/utils/extractb0/main'
include { TOPUP_EDDY } from '../topup_eddy/main'


workflow PREPROC_DWI {

    take:
        ch_dwi           // channel: [ val(meta), dwi, bval, bvec ]
        ch_rev_dwi       // channel: [ val(meta), rev-dwi, bval, bvec ], optional
        ch_b0            // Channel: [ val(meta), b0 ], optional
        ch_rev_b0        // channel: [ val(meta), rev-b0 ], optional
        ch_config_topup  // channel: [ 'topup.cnf' ], optional

    main:

        ch_versions = Channel.empty()

        ch_dwi_bvalbvec = ch_dwi
            .multiMap { meta, dwi, bval, bvec ->
                dwi:    [ meta, dwi ]
                bvs_files: [ meta, bval, bvec ]
            }

        ch_rev_dwi_bvalbvec = ch_rev_dwi
            .multiMap { meta, dwi, bval, bvec ->
                rev_dwi:    [ meta, dwi ]
                rev_bvs_files: [ meta, bval, bvec ]
            }

        // ** Denoise DWI ** //
        ch_denoise_dwi = ch_dwi_bvalbvec.dwi
            .map{ it + [[]] }

        DENOISE_DWI ( ch_denoise_dwi )
        ch_versions = ch_versions.mix(DENOISE_DWI.out.versions.first())

        // ** Denoise REV-DWI ** //
        // Need to append "rev" to the ID, to ensure output filenames
        // are different from the DWI and prevent file collisions
        //  - "cache: meta" is used to save the "real" metadata with valid ID for
        //           join operations, so it can be recovered after execution
        ch_denoise_rev_dwi = ch_rev_dwi_bvalbvec.rev_dwi
            .map{ meta, dwi -> [ [id: "${meta.id}_rev", cache: meta], dwi, [] ] }

        DENOISE_REVDWI ( ch_denoise_rev_dwi )
        ch_versions = ch_versions.mix(DENOISE_REVDWI.out.versions.first())

        // ** Eddy Topup ** //
        ch_topup_eddy_dwi = DENOISE_DWI.out.image
            .join(ch_dwi_bvalbvec.bvs_files)

        // Recover the "real" ID from "meta[cache]" (see above), to join with the bval/bvec
        ch_topup_eddy_rev_dwi = DENOISE_REVDWI.out.image
            .map{ meta, dwi -> [ meta.cache, dwi ] }
            .join(ch_rev_dwi_bvalbvec.rev_bvs_files)

        TOPUP_EDDY ( ch_topup_eddy_dwi, ch_b0, ch_topup_eddy_rev_dwi, ch_rev_b0, ch_config_topup )
        ch_versions = ch_versions.mix(TOPUP_EDDY.out.versions.first())

        // ** Bet-crop DWI ** //
        ch_betcrop_dwi = TOPUP_EDDY.out.dwi
            .join(TOPUP_EDDY.out.bval)
            .join(TOPUP_EDDY.out.bvec)

        BETCROP_FSLBETCROP ( ch_betcrop_dwi )
        ch_versions = ch_versions.mix(BETCROP_FSLBETCROP.out.versions.first())

        // ** Crop b0 ** //
        ch_crop_b0 = TOPUP_EDDY.out.b0
            .join(BETCROP_FSLBETCROP.out.bbox)

        IMAGE_CROPVOLUME ( ch_crop_b0 )
        ch_versions = ch_versions.mix(IMAGE_CROPVOLUME.out.versions.first())

        // ** N4 DWI ** //
        ch_N4 = BETCROP_FSLBETCROP.out.image
            .join(IMAGE_CROPVOLUME.out.image)
            .join(BETCROP_FSLBETCROP.out.mask)

        N4_DWI ( ch_N4 )
        ch_versions = ch_versions.mix(N4_DWI.out.versions.first())

        // ** Normalize DWI ** //
        ch_normalize = N4_DWI.out.image
            .join(TOPUP_EDDY.out.bval)
            .join(TOPUP_EDDY.out.bvec)
            .join(BETCROP_FSLBETCROP.out.mask)

        NORMALIZE_DWI ( ch_normalize )
        ch_versions = ch_versions.mix(NORMALIZE_DWI.out.versions.first())

        // ** Resample DWI ** //
        ch_resample_dwi = NORMALIZE_DWI.out.dwi
            .map{ it + [[]] }

        RESAMPLE_DWI ( ch_resample_dwi )
        ch_versions = ch_versions.mix(RESAMPLE_DWI.out.versions.first())

        // ** Extract b0 ** //
        ch_dwi_extract_b0 =   RESAMPLE_DWI.out.image
            .join(TOPUP_EDDY.out.bval)
            .join(TOPUP_EDDY.out.bvec)

        EXTRACTB0_RESAMPLE { ch_dwi_extract_b0 }
        ch_versions = ch_versions.mix(EXTRACTB0_RESAMPLE.out.versions.first())

        // ** Resample mask ** //
        ch_resample_mask = BETCROP_FSLBETCROP.out.mask
            .map{ it + [[]] }

        RESAMPLE_MASK ( ch_resample_mask )
        ch_versions = ch_versions.mix(RESAMPLE_MASK.out.versions.first())

    emit:
        dwi_resample        = RESAMPLE_DWI.out.image        // channel: [ val(meta), dwi-resampled ]
        bval                = TOPUP_EDDY.out.bval           // channel: [ val(meta), bval-corrected ]
        bvec                = TOPUP_EDDY.out.bvec           // channel: [ val(meta), bvec-corrected ]
        b0                  = EXTRACTB0_RESAMPLE.out.b0     // channel: [ val(meta), b0-resampled ]
        b0_mask             = RESAMPLE_MASK.out.image       // channel: [ val(meta), b0-mask ]
        dwi_bounding_box    = BETCROP_FSLBETCROP.out.bbox   // channel: [ val(meta), dwi-bounding-box ]
        dwi_topup_eddy      = TOPUP_EDDY.out.dwi            // channel: [ val(meta), dwi-after-topup-eddy ]
        dwi_n4              = N4_DWI.out.image              // channel: [ val(meta), dwi-after-n4 ]
        versions            = ch_versions                   // channel: [ versions.yml ]
}
