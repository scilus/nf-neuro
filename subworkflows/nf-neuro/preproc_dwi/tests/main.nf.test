nextflow_workflow {

    name "Test Subworkflow PREPROC_DWI"
    script "../main.nf"
    workflow "PREPROC_DWI"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/preproc_dwi"
    tag "subworkflows/topup_eddy"
    tag "betcrop/cropvolume"
    tag "betcrop/fslbetcrop"
    tag "denoising/mppca"
    tag "image/resample"
    tag "preproc/n4"
    tag "preproc/normalize"
    tag "utils/extractb0"
    tag "topup_eddy"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "topup_eddy_light.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("preproc_dwi_rev_b0") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec")
                    ]}
                input[1] = []
                input[2] = []
                input[3] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-PA_sbref.nii.gz")
                    ]}
                input[4] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bval,
                    workflow.out.versions,
                    file(workflow.out.b0.get(0).get(1)).name,
                    file(workflow.out.b0_mask.get(0).get(1)).name,
                    file(workflow.out.bvec.get(0).get(1)).name,
                    file(workflow.out.dwi_bounding_box.get(0).get(1)).name,
                    file(workflow.out.dwi_n4.get(0).get(1)).name,
                    file(workflow.out.dwi_resample.get(0).get(1)).name,
                    file(workflow.out.dwi_topup_eddy.get(0).get(1)).name).match() }
            )
        }
    }

    test("preproc_dwi_rev_dwi_nob0") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.nii.gz"),
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.bval"),
                        file("\${test_data_directory}/sub-01_dir-PA_dwi.bvec")
                    ]}
                input[2] = []
                input[3] = []
                input[4] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.bval,
                    workflow.out.versions,
                    file(workflow.out.b0.get(0).get(1)).name,
                    file(workflow.out.b0_mask.get(0).get(1)).name,
                    file(workflow.out.bvec.get(0).get(1)).name,
                    file(workflow.out.dwi_bounding_box.get(0).get(1)).name,
                    file(workflow.out.dwi_n4.get(0).get(1)).name,
                    file(workflow.out.dwi_resample.get(0).get(1)).name,
                    file(workflow.out.dwi_topup_eddy.get(0).get(1)).name).match() }
            )
        }
    }
}
