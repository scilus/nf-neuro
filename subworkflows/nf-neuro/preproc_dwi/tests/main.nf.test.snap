{
    "preproc_dwi_rev_dwi_nob0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,4c61c53078316c31b4d5daf446a3d6ac"
                ]
            ],
            [
                "versions.yml:md5,0280006ceecc575513278292734ffb27",
                "versions.yml:md5,0b8c908e52917b0b706fc9d1b4d6cd24",
                "versions.yml:md5,0e680a0c6b56892e1a7ec7f85bb95322",
                "versions.yml:md5,3340c4b5e56ca52f7342be3c9044eb47",
                "versions.yml:md5,7ddf2e98f59b19c9b933670550f26ad7",
                "versions.yml:md5,b34891b6b1ba869be33799b2904aad99",
                "versions.yml:md5,c15ba5efd24564dba4710b6da8c4b791",
                "versions.yml:md5,d3010f785c88ee5946806a8a758e7efb",
                "versions.yml:md5,d68fe3399120f84f1add62181708780b",
                "versions.yml:md5,ea5858879452a59bb355228ae7f38111",
                "versions.yml:md5,f041502e22449973d84ac1c618e8ebf9"
            ],
            "test_b0.nii.gz",
            "test_resampled.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test__image_boundingBox.pkl",
            "test__image_n4.nii.gz",
            "test_resampled.nii.gz",
            "test__dwi_corrected.nii.gz"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-01T17:09:44.502618804"
    },
    "preproc_dwi_rev_b0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,0280006ceecc575513278292734ffb27",
                "versions.yml:md5,0b8c908e52917b0b706fc9d1b4d6cd24",
                "versions.yml:md5,0e680a0c6b56892e1a7ec7f85bb95322",
                "versions.yml:md5,3340c4b5e56ca52f7342be3c9044eb47",
                "versions.yml:md5,7ddf2e98f59b19c9b933670550f26ad7",
                "versions.yml:md5,b34891b6b1ba869be33799b2904aad99",
                "versions.yml:md5,c15ba5efd24564dba4710b6da8c4b791",
                "versions.yml:md5,d3010f785c88ee5946806a8a758e7efb",
                "versions.yml:md5,ea5858879452a59bb355228ae7f38111",
                "versions.yml:md5,f041502e22449973d84ac1c618e8ebf9"
            ],
            "test_b0.nii.gz",
            "test_resampled.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test__image_boundingBox.pkl",
            "test__image_n4.nii.gz",
            "test_resampled.nii.gz",
            "test__dwi_corrected.nii.gz"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-01T17:04:12.699334691"
    }
}